File: 2015/python/day22.py
import enum
import fileinput

from utils.graphs import bfs


class Spell(enum.Enum):
    MAGIC_MISSILE = 0
    DRAIN = 1
    SHIELD = 2
    POISON = 3
    RECHARGE = 4

    def cost(self):
        return SPELL_COSTS[self]


SPELL_COSTS = {
    Spell.MAGIC_MISSILE: 53,
    Spell.DRAIN: 73,
    Spell.SHIELD: 113,
    Spell.POISON: 173,
    Spell.RECHARGE: 229,
}
MIN_SPELL_COST = min(SPELL_COSTS.values())
EFFECT_DURATIONS = {
    Spell.SHIELD: 6,
    Spell.POISON: 6,
    Spell.RECHARGE: 5,
}


STARTING_HP = 50
STARTING_MP = 500


class State:
    def __init__(self, hp, mp, boss_hp, boss_damage, tick_spell, mana_spent, effects):
        self.hp = hp
        self.mp = mp
        self.boss_hp = boss_hp
        self.boss_damage = boss_damage
        self.effects = effects
        self.mana_spent = mana_spent
        self.tick_spell = tick_spell

    def valid_spells(self):
        # Can recast spells if they'll wear out this turn
        return set(
            spell for spell in Spell
            if self.effects.get(spell, 0) <= 1 and spell.cost() < self.mp
        )

    def process_effects(self):
        for effect in list(self.effects.keys()):
            if effect == Spell.POISON:
                self.boss_hp -= 3
            elif effect == Spell.RECHARGE:
                self.mp += 101
            self.effects[effect] -= 1
            if self.effects[effect] == 0:
                del self.effects[effect]

    def tick(self):
        # Player turn
        self.process_effects()
        if self.boss_hp <= 0:
            return self
        if self.mp < MIN_SPELL_COST:
            self.hp = 0
            return self

        self.mp = self.mp - self.tick_spell.cost()
        self.mana_spent += self.tick_spell.cost()
        if self.tick_spell == Spell.MAGIC_MISSILE:
            self.boss_hp -= 4
        elif self.tick_spell == Spell.DRAIN:
            self.boss_hp -= 2
            self.hp += 2
        else:
            self.effects[self.tick_spell] = EFFECT_DURATIONS[self.tick_spell]

        # Boss turn
        self.process_effects()
        if self.boss_hp > 0:
            armor = 7 if Spell.SHIELD in self.effects else 0
            self.hp -= max(self.boss_damage - armor, 1)

    def next_state(self, spell, hardmode=False):
        new_state = State(
            self.hp,
            self.mp,
            self.boss_hp,
            self.boss_damage,
            spell,
            self.mana_spent,
            dict(self.effects),
        )
        if hardmode:
            new_state.hp -= 1
        if new_state.hp > 0:
            new_state.tick()

        return new_state

    def __repr__(self):
        return "State(%s, %s, %s, %s, %s, %s, %s)" % (
            self.hp,
            self.mp,
            self.boss_hp,
            self.boss_damage,
            self.tick_spell,
            self.mana_spent,
            self.effects,
        )


def run_fight(hp, mp, boss_hp, boss_damage, hardmode=False, debug=False):
    starting_state = State(hp, mp, boss_hp, boss_damage, None, 0, {})

    def neighbors(state):
        if state.hp > 0 and state.boss_hp > 0:
            for spell in state.valid_spells():
                yield state.next_state(spell, hardmode)

    for node in bfs(starting_state, neighbors, lambda s: s.mana_spent):
        state = node.value
        if state.hp > 0 and state.boss_hp <= 0:
            if debug:
                for s in reversed(list(node.get_path())):
                    print(s)
            return state


def part1(boss_hp, boss_damage):
    final_state = run_fight(STARTING_HP, STARTING_MP, boss_hp, boss_damage)

    return final_state.mana_spent


def part2(boss_hp, boss_damage):
    final_state = run_fight(
        STARTING_HP, STARTING_MP, boss_hp, boss_damage, hardmode=True
    )

    return final_state.mana_spent


if __name__ == "__main__":
    lines = [line.strip() for line in fileinput.input()]
    boss_hp, boss_damage = (int(line.split(": ")[1]) for line in lines)

    assert run_fight(10, 250, 13, 8).mana_spent == 226
    assert run_fight(10, 250, 14, 8).mana_spent == 641
    print("All tests passed")

    print("Part 1: %s" % part1(boss_hp, boss_damage))
    print("Part 2: %s" % part2(boss_hp, boss_damage))

File: 2015/rust/src/bin/day22.rs
#![feature(try_blocks)]

use std::collections::BinaryHeap;

use anyhow::{anyhow, Result};
use strum::IntoEnumIterator;

fn main() -> Result<()> {
    let input = aoc::utils::get_input()?;
    let (boss_hp, boss_damage) = parsing::parse_boss(input.trim())?;

    println!("Part 1: {}", part1(boss_hp, boss_damage)?);
    println!("Part 2: {}", part2(boss_hp, boss_damage)?);

    Ok(())
}

fn part1(boss_hp: u32, boss_damage: u32) -> Result<u32> {
    let start = FightState::new(50, boss_hp, 500);
    let win_state = run_fight(start, boss_damage, false).ok_or(anyhow!("Failed to win fight"))?;
    Ok(win_state.mp_spent)
}

fn part2(boss_hp: u32, boss_damage: u32) -> Result<u32> {
    let start = FightState::new(50, boss_hp, 500);
    let win_state = run_fight(start, boss_damage, true).ok_or(anyhow!("Failed to win fight"))?;
    Ok(win_state.mp_spent)
}

fn run_fight(start: FightState, boss_damage: u32, hard_mode: bool) -> Option<FightState> {
    let mut heap = BinaryHeap::new();
    heap.push(start);

    while let Some(state) = heap.pop() {
        if state.hp > 0 && state.boss_hp == 0 {
            return Some(state);
        }
        for new_state in state.options(boss_damage, hard_mode) {
            heap.push(new_state);
        }
    }
    None
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct FightState {
    hp: u32,
    boss_hp: u32,
    mp: u32,
    effects: Effects,
    mp_spent: u32,
}

impl FightState {
    fn new(hp: u32, boss_hp: u32, mp: u32) -> FightState {
        FightState {
            hp,
            boss_hp,
            mp,
            effects: Effects::default(),
            mp_spent: 0,
        }
    }

    fn options(&self, boss_damage: u32, hard_mode: bool) -> impl Iterator<Item = FightState> + '_ {
        Spell::iter().filter_map(move |spell| self.tick(spell, boss_damage, hard_mode))
    }

    fn tick(&self, spell: Spell, boss_damage: u32, hard_mode: bool) -> Option<FightState> {
        let mut state = *self;
        if hard_mode {
            state.hp -= 1;
            if state.hp == 0 {
                return None;
            }
        }

        // Player Turn
        state.process_effects();
        if state.boss_hp == 0 {
            return Some(state);
        }
        if spell.cost() > state.mp || state.effects[spell] != 0 {
            return None;
        }
        state.mp -= spell.cost();
        state.mp_spent += spell.cost();
        match spell {
            Spell::MagicMissile => state.boss_hp = state.boss_hp.saturating_sub(4),
            Spell::Drain => {
                state.boss_hp = state.boss_hp.saturating_sub(2);
                state.hp += 2
            }
            spell => state.effects[spell] = spell.duration(),
        }

        // Boss Turn
        state.process_effects();
        if state.boss_hp == 0 {
            return Some(state);
        }
        let shielded = state.effects[Spell::Shield] > 0;
        let armor = if shielded { 7 } else { 0 };
        let damage = boss_damage.saturating_sub(armor).max(1);
        state.hp = state.hp.saturating_sub(damage);
        if state.hp == 0 {
            return None;
        }

        Some(state)
    }

    fn process_effects(&mut self) {
        if self.effects[Spell::Poison] > 0 {
            self.boss_hp = self.boss_hp.saturating_sub(3);
        }
        if self.effects[Spell::Recharge] > 0 {
            self.mp += 101;
        }
        for spell in Spell::iter() {
            self.effects[spell] = self.effects[spell].saturating_sub(1);
        }
    }
}

impl std::cmp::PartialOrd for FightState {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl std::cmp::Ord for FightState {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        other.mp_spent.cmp(&self.mp_spent)
    }
}

#[derive(Debug, Clone, Copy, strum_macros::EnumIter)]
enum Spell {
    MagicMissile,
    Drain,
    Shield,
    Poison,
    Recharge,
}

impl Spell {
    const COSTS: [u32; 5] = [53, 73, 113, 173, 229];
    const DURATIONS: [u8; 5] = [0, 0, 6, 6, 5];

    fn cost(&self) -> u32 {
        Spell::COSTS[*self as usize]
    }

    fn duration(&self) -> u8 {
        Spell::DURATIONS[*self as usize]
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default)]
struct Effects([u8; 5]);

impl std::ops::Index<Spell> for Effects {
    type Output = u8;

    fn index(&self, index: Spell) -> &Self::Output {
        &self.0[index as usize]
    }
}

impl std::ops::IndexMut<Spell> for Effects {
    fn index_mut(&mut self, index: Spell) -> &mut Self::Output {
        &mut self.0[index as usize]
    }
}

mod parsing {
    pub fn parse_boss(s: &str) -> anyhow::Result<(u32, u32)> {
        let parts: Option<_> = try {
            let (hp, damage) = s.split_once('\n')?;
            let hp = hp.strip_prefix("Hit Points: ")?;
            let damage = damage.strip_prefix("Damage: ")?;
            (hp, damage)
        };
        let (hp, damage) = parts.ok_or_else(|| anyhow::anyhow!("Invalid boss\n{}.", s))?;
        Ok((hp.parse()?, damage.parse()?))
    }
}

#[cfg(test)]
mod tests {
    use super::{run_fight, Effects, FightState};

    #[test]
    fn fight1() {
        let start = FightState::new(10, 13, 250);
        let boss_damage = 8;
        let result = run_fight(start, boss_damage, false);

        let expected = FightState {
            hp: 2,
            boss_hp: 0,
            mp: 24,
            effects: Effects([0, 0, 0, 3, 0]),
            mp_spent: 226,
        };

        assert_eq!(result, Some(expected));
    }

    #[test]
    fn fight2() {
        let start = FightState::new(10, 14, 250);
        let boss_damage = 8;
        let result = run_fight(start, boss_damage, false);

        let expected = FightState {
            hp: 1,
            boss_hp: 0,
            mp: 114,
            effects: Effects([0, 0, 0, 3, 0]),
            mp_spent: 641,
        };

        assert_eq!(result, Some(expected));
    }
}
